### ISSUE TITLE
Named layer injection for Scala 3

### ISSUE THREAD

**Issue by 987Nabil** [zio/zio#7338]
Hey!
I thought about a way to provide named injection for Scala 3 and came up with the following

```scala
import zio.*

opaque type WithName[+A, Name <: String] <: A = A

object WithName:

  def apply[Name <: String]: [A] => A => WithName[A, Name] = [A] => (a: A) => create[Name, A](a)
  def create[Name <: String, A](a: A): WithName[A, Name]   = a

object WithNameTest extends ZIOAppDefault:

  val a: String WithName "a" = WithName["a"]("A")
  val b: String WithName "b" = WithName["b"]("B")

  override def run =
    (for
      aStr <- ZIO.service[String WithName "a"]
      bStr <- ZIO.service[String WithName "b"]
      _    <- Console.printLine(s"a: $aStr, b: $bStr")
    yield ())
      .provide(ZLayer.succeed(a), ZLayer.succeed(b))
```

This code compiles with ZIO 2.0.0-RC5, but does not work, since ZIO/Scala thinks a and b are having the same type.
It does not compile with newer ZIO versions, which I guess is related to #6829 
Idk if it is possible to make this work. Maybe with some custom `Tag` impl. ? Because tags are the keys in the map of provided dependencies right?

**Comment by adamgfraser** [comment]
It may be easier to do this with `serviceAt`. But yes I think resolving the issue around opaque types in general would allow you to do this if you want.

**Comment by 987Nabil** [comment]
I actually did not know about `serviceAt`. And I remember in a training or symposium the answer for how to inject two objects of the same type was a wrapper class. But I found the right [paragraph in the docs](https://zio.dev/reference/contextual/zenvironment/#providing-multiple-instance-of-the-same-interface) 🙂 
Still I think this way looks more elegant and requires less code on the user side. But I know this is only possible with Scala 3.

**Comment by adamgfraser** [comment]
Yeah I think that should be possible in user land once the more general issue with opaque types is resolved.

**Comment by neko-kai** [comment]
Related https://github.com/zio/zio/issues/7338

**Comment by hearnadam** [comment]
ZIO 3?
